### DISCUSSION TITLE
Nesting menuAddOn in a PropositionSet to signal restrictions

### DISCUSSION THREAD

**Discussion by emreaydinceren** [schemaorg/schemaorg#2981] [Ideas]
A PropositionSet container with maximumCount and minimumCount properties for add-ons reflects real world configuration where restricting user to limit choices in a group from 0 to N rather than signaling anything in the collection is available for choice.

For example:  
```
 "hasMenuItem":{
            "@type":"MenuItem",
            "name":"Aloo Gobi Taco",
            "description":"Mexico City-style street corn tortilla taco filled with a flavorful mixture of mildly south Indian spiced cauliflower, potato, tomato, onions and bell peppers.",
            "offers":{
               "@type":"Offer",
               "price":"3.50",
               "priceCurrency":"USD"
            },
            "nutrition":{
               "@type":"NutritionInformation",
               "calories":"170 calories",
               "fatContent":"3 grams",
               "fiberContent":"2 grams",
               "proteinContent":"4 grams"
            },
            "suitableForDiet":"https://schema.org/GlutenFreeDiet",
            "menuAddOn": [ 
            {  "@type":"PropositionSet",
                "name":"Salsa",
                "description":"Choose one salsa",
                "maximumCount": 1,
                "minimumCount: 1,
                "hasMenuItem": [ { ..hot.. }, { ..mild.. }]
                },
            {  "@type":"PropositionSet",
                    "name":"AddOns",
                    "description":"Add Beef or Chicken",
                    "minimumCount: 0,
                    "hasMenuItem": [ { ..Beef.. }, { ..Chicken.. }]
                },                
            ] 
         }
```

Above hot or mild salsa is a mutually exclusive choice whereas Beef or Chicken add-ons are optional with no upper limit.

**Comment by thadguidry** [comment]
Interesting!  What if we simplified this and made a new intangible for exclusive OR so that some properties could potentially ask a question to narrow an array of Choice? We might just simply frame those up as new properties called, OneOf and AnyOf for easier translations?  This is also the wording used in many search filtering mechanisms and their APIs behind the scenes as well as in programming languages.

StructuredValue and PropertyValue cannot present a question.

 What do you think @RichardWallis ?

**Reply by saumier** [reply]
Just a quick comment.. but my gut feeling is that this type of constraint (min max count, exclusive OR) would better expressed in an accompanying SHACL file.

**Reply by thadguidry** [reply]
@saumier agree, but SHACL would have to be an expected Type for a constraint intangible property.  It would be nicer to give publishers who might be novice or not need to have a deep ruleset capability.  Consumers also then can more easily understand these simple logic statements without much work effort on parsing and filtering accommodations.  Of course we can always incorporate a constraint intangible if we can all come to some consensus on how that might look like.  And consumers could use validators such as https://github.com/mulesoft-labs/json-ld-schema etc.  Current SHACL validators are a bit slow but hopefully others put effort into improving their performance.
